hedera-node/hedera-network-admin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/ReadableFreezeUpgradeActions.java

```diff
@@ -179,6 +201,87 @@ private void extractAndReplaceArtifacts(
         }
     }
 
+    private void generateConfigPem(@NonNull final Path artifactsLoc) {
+        requireNonNull(artifactsLoc, "Cannot generate config.txt without a valid artifacts location");
+
+        final var configTxt = artifactsLoc.resolve("config.txt");
+        final var nodeState = nodeStore.nodesState();
+
+        if (nodeState == null || nodeState.size() == 0) {
+            log.info("Node state is empty, cannot generate config.txt"); // change to log error later
+            return;
+        }
+
+        final var nodeIds = nodeStore.nodesState().keys();
+
+        try (FileWriter fw = new FileWriter(configTxt.toFile());
+                BufferedWriter bw = new BufferedWriter(fw)) {
+            Streams.stream(nodeIds)
+                    .map(nodeId -> nodeStore.get(nodeId.number()))
+                    .filter(node -> node != null && !node.deleted())
```

Comment:
node can't be null here right ?

Reply:
Since ReadableNodeStore.get(long) is Nullable, added check here to be sure.

Reply:
But you are iterating through the keys that you got from state right ? If this is the case, isn't that a catastrophic error ?